### ISSUE TITLE
[Bug]: tkAgg (and others) event look issue

### ISSUE THREAD

**Issue by 2sn** [matplotlib/matplotlib#26714]
### Bug summary

tk hangs/crashes if windows are closed not in reverse order of opened.  This happens in ipython shell.  

I have no other issues with ipython, only with matplotlib.

### Code for reproduction

```python
%matplotlib tk
figure()
from matplotlib.pylab import *
ion()
plot([0,1])
figure()
plot([0,1])

# Then close Window 1.
```

### Actual outcome

Window 2 becomes non-responsive.

### Expected outcome

There should be no crash irrespective in what order windows are closed.  

In some cases the shell hangs entirely, and the terminal window has to be closed (`^C` or `^Z` do not work).

### Additional information

It happens 100% of the time.

### Operating system

Fedora 38

### Matplotlib Version

3.7.2

### Matplotlib Backend

TkAgg

### Python version

3.11.4 (build from source)

### Jupyter version

8.15.0 (installed from pip)

### Installation

pip

**Comment by 2sn** [comment]
https://github.com/ipython/ipython/issues/14155

**Comment by anntzer** [comment]
I can repro the issue on macos; it only occurs in ipython (8.14.0, haven't tried older ones), not in a plain python shell, and goes at least as far back as matplotlib 3.0.3.

**Comment by 2sn** [comment]
It may be present since  prompt-toolkit! >=3 is used in IPython.  Yet I do not know whether the issue is in IPython or in Matplotlib - they just seem to be not happy together.

**Comment by tacaswell** [comment]
My guess is that the issue is in https://github.com/ipython/ipython/blob/main/IPython/terminal/pt_inputhooks/tk.py

We go through some jigger-pokery to make our mainloop logic work (https://github.com/matplotlib/matplotlib/blob/e9f1d7b984488771eff0403585ecb75c0697cf7e/lib/matplotlib/backends/_backend_tk.py#L513-L525) but I do not remember all of the details of why we do it that may.

My guess as to what is going on here is that the input hook is waiting on the "wrong" root and the keyboard events that should be going through it never come in to exit the event loop and give control back to the prompt. 

I suspect the signals do not work because the signal handlers that Python installs note they were fired and then set a flag that will be handled the next time that the byte code processing loop runs.  However if the event loop is running, from Python's point of view we are in a c-library call and it is waiting for that to return before it handles the SIGINT!  We have a bunch of logic in the Qt backend (https://github.com/matplotlib/matplotlib/blob/e9f1d7b984488771eff0403585ecb75c0697cf7e/lib/matplotlib/backends/qt_compat.py#L165-L230) to make this work in the Qt case.  We may need something similar for Tk?

attn @richardsheridan and @daniilS who understand Tk :)

**Comment by 2sn** [comment]
As a note, the issue seems to be also not present in the `gtk4` backend.

**Comment by richardsheridan** [comment]
> My guess is that the issue is in https://github.com/ipython/ipython/blob/main/IPython/terminal/pt_inputhooks/tk.py

I agree with this. It seems to be an unfortunate consequence of `tkinter` conflating windows with the tcl interpreter and IPython needing to rely on its internal global state.  I don't know how to fix it but I think there's nothing to be done here.
